### ISSUE TITLE
[Task] Separate `DID` from `DIDUrl`

### ISSUE THREAD

**Issue by cycraig** [iotaledger/identity.rs#408]
## Description
Split `CoreDID` and `IotaDID` into separate `DID` and `DIDUrl` structs.

E.g.:

- `DID`: `"did:iota:main:9rK6DPF46MCEzgfLD8AHFsaTuMqvmRo6kbXfjqQJPJmC"`
- `DIDUrl`: `"did:iota:main:9rK6DPF46MCEzgfLD8AHFsaTuMqvmRo6kbXfjqQJPJmC#keys?query/path`

![image](https://user-images.githubusercontent.com/5772887/133995369-4e51b598-32fd-4849-9a98-94039d9a4703.png)

## Motivation
We currently have two `DID` structs: `CoreDID` and `IotaDID`:

- `CoreDID` is a re-export of the [`did-url` crate](https://crates.io/crates/did_url), which is optimised to  represent a full `DIDUrl`, potentially including a fragment, query, and/or path.
- `IotaDID` just wraps `CoreDID` to enforce the [IOTA DID Method specification](https://identity.docs.iota.org/docs/specs/iota_did_method_spec).

The discussion in #387 makes apparent that we need some way to distinguish a `DID` from a `DIDUrl`. Specifically to enforce invariants in structs and improve clarity in function parameters.

For example:

- a `CoreDocument` uses `CoreDID` (`DIDUrl`) for the `id` and `controller` fields, which should [conform to the base `DID` syntax according to the did-core specification](https://www.w3.org/TR/did-core/#did-document-properties), but currently theoretically allows any `DIDUrl`.
- `Client::resolve` takes an `IotaDID` (`DIDUrl`) parameter but ignores any URL segments and just uses the base `DID`.
- the `Account` defines an `IdentityTag` which represents just the network and tag of a `DID` for mapping internally.

Note that any solution should translate well to the external Wasm bindings API, which does not support generics or interfaces properly.

NOTE: any invariants should be enforced when deserialising these structures.

## Resources
`DID`: https://www.w3.org/TR/did-core/#did-syntax
`DIDUrl`: https://www.w3.org/TR/did-core/#did-url-syntax

## To-do list
Please link PRs that match the To-do list item behind the item after it has been submitted.

- [ ] Implement `CoreDID` and `CoreDIDUrl`
- [ ] Implement `IotaDID` and `IotaDIDUrl`
- [ ] Replace `IdentityTag`, `IdentityKey` with `IotaDID` in the `Account`

## Change checklist
Add an `x` to the boxes that are relevant to your changes, and delete any items that are not.

- [ ] The feature is implemented in Rust and across all bindings whereas possible.
- [ ] The feature has sufficient testing coverage
- [ ] All tests and examples build and run locally as expected
- [ ] Every piece of code has been document according to the documentation guidelines.
- [ ] If conceptual documentation (mdbook) and examples highlighting the feature exist, they are properly updated.
- [ ] If the feature is not currently documented, a documentation task Issue has been opened to address this.

**Comment by PhilippGackstatter** [comment]
Very nice proposal!

Does that mean that `Account::sign` would in the future take just a `IotaDidUrl`? E.g.:

```rust
BEFORE: account.sign(did, "key-1", &mut credential).await?;
AFTER: account.sign(did_url, &mut credential).await?;
```

and I imagine it would be nice to have methods that can easily convert to a url. For example

```rust
let did: IotaDid = "did:iota:main:9rK6DPF46MCEzgfLD8AHFsaTuMqvmRo6kbXfjqQJPJmC".parse().unwrap();
let did_url: IotaDidUrl = did.to_url().set_fragment("key-1");
account.sign(did_url, &mut credential).await?;
```

Seems like that would be a nice-to-have to make working with `Account::sign` easy, given just an `IotaDid`. It can look different, of course, but the point is, it should be easy to convert.
